### ISSUE TITLE
support for Ed25519

### ISSUE THREAD

**Issue by henrikalves** [lbuchs/WebAuthn#78]
On the current draft of the W3C it is stated that:

Relying Parties that wish to support a wide range of authenticators **SHOULD include at least the following** COSEAlgorithmIdentifier values:

- -8 (Ed25519)
- -7 (ES256)
- -257 (RS256)

https://w3c.github.io/webauthn/#dom-publickeycredentialcreationoptions-pubkeycredparams

But a draft is just a draft so i was wondering if its a good idea to add support for Ed25519

**Comment by paragonie-security** [comment]
> But a draft is just a draft so i was wondering if its a good idea to add support for Ed25519

Ed25519 is the most secure of the three options provided in that list.

**Comment by lbuchs** [comment]
I've already a implementation under https://github.com/lbuchs/WebAuthn/tree/feature/ed25519 but it's not yet tested.

**Comment by lbuchs** [comment]
Since v3.0.1 OpenSSL CLI supports Ed25519 (see [here](https://github.com/openssl/openssl/issues/6988)), but unfortunately it's not yet supported by PHP.

small test script:
```php
<?php
echo "openssl version: " . OPENSSL_VERSION_TEXT . "<br />\n";

$data = 'hello world';
$publicKey  = '-----BEGIN PUBLIC KEY-----' . "\r\n";
$publicKey .= 'MCowBQYDK2VwAyEAm8aIHTZPVRhOEG44ewdrG5MH68kbw5lZt7XQpUOaMo4=' . "\r\n";
$publicKey .= '-----END PUBLIC KEY-----'  . "\r\n";

$signature = base64_decode('IYOFlr2B/HdXfaWtmvX3XK+VGHCtTV3Q/hCk9M45NET6xa1AHCVvm64cmCVjBSuCd5lG5UK5qM9O2yM3lFUMDg==');

openssl_verify($data, $signature, $publicKey, OPENSSL_ALGO_SHA256);

print(openssl_error_string() . '<br/><br/>');
print('supported curves: <ul><li>' . implode('</li><li>', openssl_get_curve_names()) . '</li></ul>');
```

I'm ending with:
```
openssl version: OpenSSL 3.0.8 7 Feb 2023
error:03000096:digital envelope routines::operation not supported for this keytype
```

As I want to keep this library simple I will not go for a third party library to support Ed25519, I'll wait until it's supported by PHP-OpenSSL.

**Comment by henrikalves** [comment]
yeah using the [Sodium](https://www.php.net/manual/de/book.sodium.php) extension it would work

```php
<?php
if(!extension_loaded('Sodium')) {
	die("Sodium not installed");
}

$data = 'hello world';
$signature = base64_decode('IYOFlr2B/HdXfaWtmvX3XK+VGHCtTV3Q/hCk9M45NET6xa1AHCVvm64cmCVjBSuCd5lG5UK5qM9O2yM3lFUMDg==');

$publickey = base64_decode("MCowBQYDK2VwAyEAm8aIHTZPVRhOEG44ewdrG5MH68kbw5lZt7XQpUOaMo4=");
$publickey = substr($publickey, strlen($publickey) - 32);

// Verify the signature using Sodium
$isValid = sodium_crypto_sign_verify_detached($signature, $data, $publickey);

if ($isValid) {
	echo "The signature is valid.";
} else
{
	echo "The signature is NOT valid.";
}
```

how i am getting the 32 byte key is probably very stupid btw.

there is also an php polyfill by paragonie [sodium_compat](https://github.com/paragonie/sodium_compat) for people that dont have the extension installed.

But yeah if you dont want to use a "third party library" then this will have to wait sadly

**Comment by lbuchs** [comment]
thanks for that, interesting. I'll implement it with Sodium / Sodium Compat ( check with `function_exist('sodium_crypto_sign_verify_detached')` )

**Comment by lbuchs** [comment]
implemented with 13b68435e20c40ed17f7242cf733db5508862649
